pkg/frontend/querymiddleware/astmapper/sharding_test.go

```diff
@@ -542,14 +573,43 @@ func TestShardSummer(t *testing.T) {
 			out:                      `count(group without() (` + concatShards(t, shardCount, `group without() ({namespace="foo",__query_shard__="x_of_y"})`) + `))`,
 			expectedShardableQueries: 1,
 		},
+		{
+			in:                       `vector(scalar(sum(foo)))`,
+			out:                      `vector(scalar(sum(` + concatShards(t, shardCount, `sum(foo{__query_shard__="x_of_y"})`) + `)))`,
+			expectedShardableQueries: 1,
+		},
 	} {
 		t.Run(tt.in, func(t *testing.T) {
-			runTest(t, tt.in, tt.out, shardCount, false, tt.expectedShardableQueries*shardCount)
-		})
+			t.Run("applying sharding", func(t *testing.T) {
+				runTest(t, tt.in, tt.out, shardCount, false, tt.expectedShardableQueries*shardCount)
+			})
+
+			t.Run("inspecting", func(t *testing.T) {
+				// The query should be unchanged, but the number of sharded queries should still be reported.
+				runTest(t, tt.in, tt.in, 1, true, tt.expectedShardableQueries)
+			})
+
+			t.Run("checking if all vector selectors are sharded", func(t *testing.T) {
+				// There are a lot of edge cases in willShardAllSelectors, and this method is only called
+				// if an expression appears inside a binary expression.
+				// So rather than generating a whole set of test cases just for this, we directly check each of
+				// our existing test cases.
+				expr, err := parser.ParseExpr(tt.in)
+				require.NoError(t, err)
+				stats := NewMapperStats()
+				summer := newShardSummer(shardCount, false, EmbeddedQueriesSquasher, log.NewNopLogger(), stats, newQueryShardLabeller(shardCount))
+
+				willShardAllSelectors, err := summer.willShardAllSelectors(expr)
+				require.NoError(t, err)
+				hasVectorSelectors, err := AnyNode(expr, isVectorSelector)
+				require.NoError(t, err)
 
-		t.Run(tt.in+" with 'inspect only' set", func(t *testing.T) {
-			// The query should be unchanged, but the number of sharded queries should still be reported.
-			runTest(t, tt.in, tt.in, 1, true, tt.expectedShardableQueries)
+				if hasVectorSelectors {
+					require.Equal(t, tt.expectedShardableQueries > 0, willShardAllSelectors)
```

Comment:
Can you add a message here since the default failure message of "true is not equal to false" isn't very helpful?

Reply:
Good idea, added in 2ac504d